From 96752a8b1c9bb59159fec294a6db8a1f2787df63 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:21 -0400
Subject: [fs] btrfs: retrieve more info from FS_INFO ioctl

Message-id: <1407179109-13714-315-git-send-email-sandeen@redhat.com>
Patchwork-id: 87223
O-Subject: [RHEL7.1 PATCH 314/422] btrfs: retrieve more info from FS_INFO ioctl
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: David Sterba <dsterba@suse.cz>

commit 80a773fbfc2d6b5b2478377e8ac271d495f55e73
Author: David Sterba <dsterba@suse.cz>
Date:   Wed May 7 18:17:06 2014 +0200

    btrfs: retrieve more info from FS_INFO ioctl

    Provide the basic information about filesystem through the ioctl:
    * b-tree node size (same as leaf size)
    * sector size
    * expected alignment of CLONE_RANGE and EXTENT_SAME ioctl arguments

    Backward compatibility: if the values are 0, kernel does not provide
    this information, the applications should ignore them.

    Signed-off-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/ioctl.c b/fs/btrfs/ioctl.c
index 2f6d7b1..c6c8e35 100644
--- a/fs/btrfs/ioctl.c
+++ b/fs/btrfs/ioctl.c
@@ -2574,6 +2574,10 @@ static long btrfs_ioctl_fs_info(struct btrfs_root *root, void __user *arg)
  }
  mutex_unlock(&fs_devices->device_list_mutex);
 
+ fi_args->nodesize = root->fs_info->super_copy->nodesize;
+ fi_args->sectorsize = root->fs_info->super_copy->sectorsize;
+ fi_args->clone_alignment = root->fs_info->super_copy->sectorsize;
+
  if (copy_to_user(arg, fi_args, sizeof(*fi_args)))
   ret = -EFAULT;
 
diff --git a/include/uapi/linux/btrfs.h b/include/uapi/linux/btrfs.h
index 901a3c5..7554fd3 100644
--- a/include/uapi/linux/btrfs.h
+++ b/include/uapi/linux/btrfs.h
@@ -181,7 +181,11 @@ struct btrfs_ioctl_fs_info_args {
  __u64 max_id;    /* out */
  __u64 num_devices;   /* out */
  __u8 fsid[BTRFS_FSID_SIZE];  /* out */
- __u64 reserved[124];   /* pad to 1k */
+ __u32 nodesize;    /* out */
+ __u32 sectorsize;   /* out */
+ __u32 clone_alignment;   /* out */
+ __u32 reserved32;
+ __u64 reserved[122];   /* pad to 1k */
 };
 
 struct btrfs_ioctl_feature_flags {
-- 
1.7.1